src/librustc/middle/const_eval.rs

```diff
@@ -1103,10 +854,28 @@ pub fn eval_const_expr_partial<'tcx>(tcx: &ty::ctxt<'tcx>,
 
           let mut call_args = NodeMap();
           for (arg, arg_expr) in decl.inputs.iter().zip(args.iter()) {
+              let arg_hint = match ty_hint {
+                  ExprTypeChecked => ExprTypeChecked,
+                  UncheckedExprNoHint | UncheckedExprHint(_) => {
+                      if let Some(hint) = tcx.ast_ty_to_ty_cache.borrow().get(&arg.ty.id) {
```

Comment:
OK, so this seems not incorrect, but using `ast_ty_to_ty_cache` outside of typeck is imo kind of dubious (I know lint does it too, that's also dubious). I've been meaning to remove it, in fact. For one thing, it doesn't play nicely with incr. comp., though I guess we could tweak it to find the appropriate item for the purposes of registering a dependency. Better would be if you called `lookup_item_type(did)` and used the resulting type from there to extract out the hints.